File: hash.rb
#!/usr/bin/env ruby
# hash.rb
# encoding: UTF-8

require 'digest'
require 'optparse'
require 'base64'

# ANSI colors
COLORS = {
  green: "\e[92m",
  red: "\e[91m",
  yellow: "\e[93m",
  blue: "\e[94m",
  reset: "\e[0m"
}

def colorize(text, color)
  "#{COLORS[color]}#{text}#{COLORS[:reset]}"
end

def compute_hash(data, algo)
  case algo
  when 'md5' then Digest::MD5.digest(data)
  when 'sha1' then Digest::SHA1.digest(data)
  when 'sha256' then Digest::SHA256.digest(data)
  when 'sha512' then Digest::SHA512.digest(data)
  else raise "Unsupported algorithm: #{algo}"
  end
end

def format_hash(digest, format)
  if format == 'hex'
    digest.unpack1('H*')
  elsif format == 'base64'
    Base64.strict_encode64(digest)
  else
    digest.unpack1('H*')
  end
end

def read_file(filename)
  File.binread(filename)
rescue => e
  raise "Cannot read file: #{e.message}"
end

def write_file(filename, content)
  File.write(filename, content, encoding: 'UTF-8')
rescue => e
  raise "Cannot write file: #{e.message}"
end

options = {
  algo: 'sha256',
  text: nil,
  input: nil,
  output: nil,
  salt: nil,
  compare: nil,
  format: 'hex',
  help: false
}

OptionParser.new do |opts|
  opts.banner = "Usage: hash.rb [options]"
  opts.on('-a', '--algo ALGO', 'Algorithm: md5, sha1, sha256, sha512 (default sha256)') { |a| options[:algo] = a }
  opts.on('-t', '--text TEXT', 'Text to hash') { |t| options[:text] = t }
  opts.on('-i', '--input FILE', 'Input file') { |f| options[:input] = f }
  opts.on('-o', '--output FILE', 'Output file') { |f| options[:output] = f }
  opts.on('-s', '--salt SALT', 'Salt string') { |s| options[:salt] = s }
  opts.on('-c', '--compare HASH', 'Hash to compare against') { |c| options[:compare] = c }
  opts.on('-f', '--format FORMAT', 'Output format: hex, base64 (default hex)') { |f| options[:format] = f }
  opts.on('-h', '--help', 'Show this help') { options[:help] = true }
end.parse!

if options[:help]
  puts <<~HELP
    Usage: hash.rb [options]
    Options:
      -a, --algo ALGO     Algorithm: md5, sha1, sha256, sha512 (default sha256)
      -t, --text TEXT     Text to hash
      -i, --input FILE    Input file
      -o, --output FILE   Output file
      -s, --salt SALT     Salt string
      -c, --compare HASH  Hash to compare against
      -f, --format FORMAT Output format: hex, base64 (default hex)
      -h, --help          Show this help
  HELP
  exit
end

# Read input
data = nil
if options[:text]
  data = options[:text].dup.force_encoding('UTF-8')
elsif options[:input]
  begin
    data = read_file(options[:input])
  rescue => e
    puts colorize("Error reading file: #{e}", :red)
    exit 1
  end
else
  # read from stdin
  if $stdin.tty?
    print colorize("Enter data (Ctrl+D to finish): ", :yellow)
  end
  data = $stdin.binread
  if data.nil? || data.empty?
    puts colorize("No input data", :red)
    exit 1
  end
end

# Add salt
if options[:salt]
  salt = options[:salt].dup.force_encoding('ASCII-8BIT')
  data = salt + data
end

# Compute hash
begin
  digest = compute_hash(data, options[:algo])
rescue => e
  puts colorize("Error: #{e.message}", :red)
  exit 1
end

result = format_hash(digest, options[:format])

# Compare
if options[:compare]
  if result == options[:compare]
    puts colorize("✅ Hash matches!", :green)
  else
    puts colorize("❌ Hash does NOT match.", :red)
    puts "Expected: #{options[:compare]}"
    puts "Computed: #{result}"
  end
  exit 0
end

# Output
if options[:output]
  begin
    write_file(options[:output], result)
    puts colorize("Hash written to #{options[:output]}", :green)
  rescue => e
    puts colorize("Error writing file: #{e}", :red)
    exit 1
  end
else
  puts result
end

File: hash.py
# hash.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import hashlib
import sys
import argparse
import base64
import os

# ANSI colors
COLORS = {
    'green': '\033[92m',
    'red': '\033[91m',
    'yellow': '\033[93m',
    'blue': '\033[94m',
    'reset': '\033[0m'
}

def colorize(text, color):
    return f"{COLORS.get(color, '')}{text}{COLORS['reset']}"

def get_hasher(algo):
    """Возвращает функцию хеширования по имени алгоритма."""
    algos = {
        'md5': hashlib.md5,
        'sha1': hashlib.sha1,
        'sha256': hashlib.sha256,
        'sha512': hashlib.sha512
    }
    if algo not in algos:
        raise ValueError(f"Unsupported algorithm: {algo}")
    return algos[algo]

def compute_hash(data, algo, salt=''):
    """Вычисляет хеш данных с солью."""
    if salt:
        data = salt + data
    if isinstance(data, str):
        data = data.encode('utf-8')
    hasher = get_hasher(algo)()
    hasher.update(data)
    return hasher.digest()

def format_hash(digest, fmt):
    if fmt == 'hex':
        return digest.hex()
    elif fmt == 'base64':
        return base64.b64encode(digest).decode('ascii')
    else:
        raise ValueError(f"Unsupported format: {fmt}")

def read_file(filename):
    with open(filename, 'rb') as f:
        return f.read()

def main():
    parser = argparse.ArgumentParser(description="HashSuite – MD5/SHA hashing tool")
    parser.add_argument('-a', '--algo', default='sha256',
                        choices=['md5', 'sha1', 'sha256', 'sha512'],
                        help='Hash algorithm (default: sha256)')
    parser.add_argument('-t', '--text', help='Text to hash')
    parser.add_argument('-i', '--input', help='Input file')
    parser.add_argument('-o', '--output', help='Output file')
    parser.add_argument('-s', '--salt', default='', help='Salt string (prepended)')
    parser.add_argument('-c', '--compare', help='Hash to compare against')
    parser.add_argument('-f', '--format', default='hex', choices=['hex', 'base64'],
                        help='Output format (default: hex)')
    parser.add_argument('-v', '--version', action='version', version='HashSuite 1.0')

    args = parser.parse_args()

    # Read input
    if args.text:
        data = args.text.encode('utf-8')
    elif args.input:
        try:
            data = read_file(args.input)
        except Exception as e:
            sys.exit(colorize(f"Error reading file: {e}", 'red'))
    else:
        # read from stdin
        if sys.stdin.isatty():
            sys.stderr.write(colorize("Enter data (Ctrl+D to finish): ", 'yellow'))
        data = sys.stdin.buffer.read()
        if not data:
            sys.exit(colorize("No input data", 'red'))

    try:
        # Compute hash
        hasher = get_hasher(args.algo)()
        if args.salt:
            hasher.update(args.salt.encode('utf-8'))
        hasher.update(data)
        digest = hasher.digest()
        result = format_hash(digest, args.format)
    except Exception as e:
        sys.exit(colorize(f"Error during hashing: {e}", 'red'))

    # Compare if requested
    if args.compare:
        if result == args.compare:
            print(colorize("✅ Hash matches!", 'green'))
        else:
            print(colorize("❌ Hash does NOT match.", 'red'))
            print(f"Expected: {args.compare}")
            print(f"Computed: {result}")
        # still output the computed hash if output specified? We'll just exit.
        if args.output:
            # but we might want to write the result anyway? Usually comparison is final.
            pass
        sys.exit(0)

    # Output
    if args.output:
        try:
            with open(args.output, 'w', encoding='utf-8') as f:
                f.write(result)
            print(colorize(f"Hash written to {args.output}", 'green'))
        except Exception as e:
            sys.exit(colorize(f"Error writing file: {e}", 'red'))
    else:
        print(result)

if __name__ == '__main__':
    main()
